### DISCUSSION TITLE
How Can I get All Nodes' refs(instance)?

### DISCUSSION THREAD

**Discussion by yczkhx-ths** [bcakmakoglu/vue-flow#1518] [Q&A] [answered]
i can use ref to get instance in slot,like ```` <MyNode ref="NodeRef">````.but when i want to get all Nodes' instance with props 'node-types', how can i get?
the case is that some functions are listed and exposed in my node file, i want to use them in vue-flow file;
````
// father.vue
<vue-flow
  class="vue-flow"
  ref="vueFlowRef"
  v-model:nodes="store.nodes"
  v-model:edges="store.edges"
  :node-types="nodesType"
>
</vue-flow>

````

````
// mynode.vue
function aExposeFunc(){
 // the function i want to use in father.vue
}
defineExpose({ aExposeFunc});
````

**Comment by yczkhx-ths** [comment]
and i see useNode provide 'nodeEl' property, but use 'nodeEl.__vnode' seems not concise

**Comment by bcakmakoglu** [comment]
> and i see useNode provide 'nodeEl' property, but use 'nodeEl.__vnode' seems not concise

Not sure what this is supposed to mean. The `nodeEl` reference is the *wrapping `<div>` element that is around all nodes*.

> but when i want to get all Nodes' instance with props 'node-types', how can i get?

You can wrap your custom node in a render function and add a ref from there.

```ts
const nodeTypes = {
	custom: (nodeProps) => h(MyNode, { ...nodeProps, ref: myRef })
}
```

Although I'd think working with a store or composable + provide/inject would be more elegant than storing refs to all nodes and then triggering functions from there but ultimately that's your choice to make ^^

**Comment by yczkhx-ths** [comment]
i tried 
````
const nodeTypes = {
	custom: (nodeProps) => h(MyNode, { ...nodeProps, ref: myRef })
}
````
it seems that i can get the newest node instance' ref.but if there are many nodes in graph, it's not work

**Reply by bcakmakoglu** [reply, marked as answer]
You can store refs in an array or a map.

```ts
const myRefs = ref(new Map())

const nodeTypes = {
	custom: (nodeProps) => h(MyNode, { ...nodeProps, ref: (nodeRef) => myRefs.value.add(nodeProps.id, nodeRef) })
}
